### ISSUE TITLE
Bug: StaticRoute returning old file contents

### ISSUE THREAD

**Issue by imbusy** [aio-libs/aiohttp#628]
I am running aiohttp in an Ubuntu Docker container within a Vagrant (VirtualBox) CentOS 7 VM within Windows 8.1. The static files are served from a mounted directory in the CentOS VM, which is again a synced folder to a Windows directory.

This setup causes old file contents to be served even though the date-modified and content-length headers are correct.

The immediate fix is not to use add_static on the router, but do this:

```
static_route = web_urldispatcher.StaticRoute(None, '/static/', '/home/web/static/')
static_route._sendfile = static_route._sendfile_fallback
app.router.register_route(static_route)
```

Bug is present in v0.18.3, but I'm not sure which part of my setup is to blame.

**Comment by asvetlov** [comment]
I suspect you've caught filesystem desynchronization  effect between Docker, Vagrant and Windows.
After tiny delay you are able to get required file content, isn't it?
P.S. You software stack is not looking production ready :)

**Comment by imbusy** [comment]
If you're talking about a delay of ~20mins, you might be right. I got the files to refresh once, but when it happened again, I investigated further. The files are fine if I "cat" them in the Docker container.

This is obviously for development :). I would not use aiohttp to serve static files in production. And that makes it an even more important bug, because files while developing change so frequently.

**Comment by asvetlov** [comment]
Sounds bad.
Well, we can add an option for switching off `sendfile` capacity even if it's available.
I cannot reproduce the issue on Linux box BTW

**Comment by jashandeep-sohi** [comment]
Mhhh...this is an odd bug.
So @imbusy, after you `cat` the file, does `aiohttp` starting serve the new content?

**Comment by imbusy** [comment]
@jashandeep-sohi , no, it still serves the old file.

**Comment by imbusy** [comment]
My guess what's happening: after saving a file, a new file may be created, but the old file handle still kept open by some process. When requesting the file handle ID, the old file handle ID is returned and the old file contents are served. This is just a guess. I am not an expert on Linux system internals.

**Comment by jashandeep-sohi** [comment]
That's what I was thinking too, but I don't think Python caches the file object returned by `open("/path/file.dat", "rb")` (or the file handle referenced by that object). `aiohttp` creates a brand new file object & thus a new file handle on every request, so I don't quite understand how it might be holding on to the old handle.

But it seems the problem even affects `ningx`, so `aiohttp` is not alone:
https://github.com/fideloper/docker-nginx-php/issues/17

**Comment by asvetlov** [comment]
It cannot be aiohttp or python issue because the code always open new file
descriptor by given filename on every file transfer.

Thus we have 2 options:
1. Leave the issue as wontfix
2. Support downgrading to not using sendfile even it is available, add to
doc a cryptic message like "If you use docker on windows the option may
help you somehow, forgot about the option otherwise".

**Comment by jashandeep-sohi** [comment]
@asvetlov I propose `StaticRoute` should fallback to `sendfile_fallback` if environment variable `AIOHTTP_SENDFILE` is set & equal to `0`.

**Comment by asvetlov** [comment]
Environment variable sounds good but I'm inclining to AIOHTTP_NOSENDFILE=x

**Comment by jashandeep-sohi** [comment]
Yea, that's better.

**Comment by asvetlov** [comment]
@jashandeep-sohi would you provide a patch?

**Comment by jashandeep-sohi** [comment]
@asvetlov Yea, tomorrow.

BTW, apparently this is a well known bug for VirtualBox, with the recommended "fix" being "turning off sendfile". I wouldn't call that "fix".
https://www.virtualbox.org/ticket/12597
https://www.virtualbox.org/ticket/9069

**Comment by asvetlov** [comment]
Yes, it's not a fix but workaround for virtualbox bug

**Comment by imbusy** [comment]
If I have voting rights, I'd go for AIOHTTP

**Comment by imbusy** [comment]
Sorry, I was on mobile and clicked something accidentally. Not sure how to delete comments.

**Comment by asvetlov** [comment]
Fixed by #629
